appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java

```diff
@@ -0,0 +1,696 @@
+     */
+    private void accept(ResourceEnvReferenceDescriptor resourceEnvRef, JndiNameEnvironment env) {
+        String jndiName = resourceEnvRef.getJndiName();
+        if (resourceEnvRef.isEJBContext() || resourceEnvRef.isValidator() || resourceEnvRef.isValidatorFactory() || resourceEnvRef.isCDIBeanManager())
+            return;
+
+        // Validate Managed Bean references now
+        String newName = convertModuleOrAppJNDIName(jndiName, env);
+        if (!myNamespace.find(newName, env)) {
+            // Every type of resource taken care of. Validate any custom resources now.
+            validateJNDIRefs(jndiName, env);
+        }
+    }
+
+    /**
+     * Convert JNDI names beginning with java:module and java:app to their corresponding java:global names.
+     *
+     * @return the converted name with java:global JNDI prefix.
+     */
+    private String convertModuleOrAppJNDIName(String jndiName, JndiNameEnvironment env) {
+        BundleDescriptor bd = null;
+        if( env instanceof EjbDescriptor ) {
+            bd = ((EjbDescriptor)env).getEjbBundleDescriptor();
+        } else if( env instanceof BundleDescriptor ) {
+            bd = (BundleDescriptor) env;
+        }
+
+        if (jndiName == null)
+            return null;
+
+        if( bd != null ) {
+            String appName = null;
+            if (!application.isVirtual()) {
+                appName = application.getAppName();
+            }
+            String moduleName = bd.getModuleDescriptor().getModuleName();
+            StringBuilder javaGlobalName = new StringBuilder("java:global/");
+            if (jndiName.startsWith(ResourceConstants.JAVA_APP_SCOPE_PREFIX)) {
+                if (appName != null) {
+                    javaGlobalName.append(appName);
+                    javaGlobalName.append("/");
+                }
+
+                // Replace java:app/ with the fully-qualified global portion
+                int javaAppLength = ResourceConstants.JAVA_APP_SCOPE_PREFIX.length();
+                javaGlobalName.append(jndiName.substring(javaAppLength));
+            }
+            else if (jndiName.startsWith(ResourceConstants.JAVA_MODULE_SCOPE_PREFIX)) {
+                if (appName != null) {
+                    javaGlobalName.append(appName);
+                    javaGlobalName.append("/");
+                }
+
+                javaGlobalName.append(moduleName);
+                javaGlobalName.append("/");
+
+                // Replace java:module/ with the fully-qualified global portion
+                int javaModuleLength = ResourceConstants.JAVA_MODULE_SCOPE_PREFIX.length();
+                javaGlobalName.append(jndiName.substring(javaModuleLength));
+            }
+            else {
+                return "";
+            }
+            return javaGlobalName.toString();
+        }
+        return "";
+    }
+
+    /**
+     * If the message destination ref is linked to a message destination, fetch the linked destination and validate it.
```

Comment:
I'm not sure what you are talking about. Didn't we realise that the problem is in tests?